File: packages/core/src/context.ts
/**
 * @file context.ts
 * @brief FileContext and ProjectContext construction for the analysis pipeline.
 *
 * Parses each source file with tree-sitter and attaches metadata that every
 * analysis pass needs: the CST, raw lines, file role, and platform hint. All
 * passes receive a `FileContext` and must not re-parse the file themselves.
 *
 * @remarks
 * Two parser singletons are kept alive for the process lifetime — one for C,
 * one for C++. tree-sitter parsers are not thread-safe, so parallel file
 * analysis must create separate parser instances or serialize access.
 *
 * @see {@link buildFileContext} for the main entry point.
 * @see {@link ProjectContext} for the cross-file shared state.
 * @module Core
 *
 * @author Nirapod Team
 * @date 2026
 *
 * SPDX-License-Identifier: APACHE-2.0
 * SPDX-FileCopyrightText: 2026 Nirapod Contributors
 */

import Parser from "tree-sitter";
import C from "tree-sitter-c";
import Cpp from "tree-sitter-cpp";
import path from "node:path";
import type { AuditConfig, FileRole, Language, PlatformHint } from "@nirapod-audit/protocol";

const cParser = new Parser();
cParser.setLanguage(C);

const cppParser = new Parser();
cppParser.setLanguage(Cpp);

/**
 * All data a single analysis pass needs to inspect one source file.
 *
 * @remarks
 * Built once per file by {@link buildFileContext} and then passed read-only
 * to every pass in the pipeline. Passes must not mutate any field.
 */
export interface FileContext {
  /** Absolute path to the source file. */
  path: string;
  /** Raw file content as a single string. */
  raw: string;
  /** File content split on `"\n"`, for line-number-based lookups. */
  lines: string[];
  /** tree-sitter parse tree (CST). */
  tree: Parser.Tree;
  /** Root node of the CST — entry point for all queries. */
  rootNode: Parser.SyntaxNode;
  /** Structural role of this file, used to gate rule subsets. */
  role: FileRole;
  /**
   * Detected or configured hardware platform.
   *
   * @remarks
   * Drives platform-specific crypto rules (CC310 vs CC312 vs ESP32).
   * Derived from `#include` guards and preprocessor macros when config is
   * `"auto"`.
   */
  platform: PlatformHint;
  /**
   * Detected source language.
   *
   * @remarks
   * Drives parser selection and determines which pass families apply to
   * this file. For example, Doxygen rules run only on `"c"` | `"cpp"` files,
   * while TSDoc rules run only on `"typescript"` files.
   */
  language: Language;
  /** `true` when the file extension is `.cpp`, `.cc`, or `.hpp`. */
  isCpp: boolean;
  /** Project-wide shared state (symbol table, config, file list). */
  project: ProjectContext;
}

/**
 * Shared state built once for an entire audit run and handed to every pass.
 *
 * @remarks
 * `definedGroups` and `definedSymbols` are populated during the pre-scan
 * phase (first pass over all headers) so that cross-file checks like
 * `@ingroup` target validation can run during the main analysis phase.
 */
export interface ProjectContext {
  /** Absolute path to the project root directory. */
  rootDir: string;
  /** Absolute paths of every file that will be analysed. */
  allFiles: string[];
  /** Active configuration for this run. */
  config: AuditConfig;
  /**
   * All `@defgroup` names declared anywhere in the project headers.
   *
   * @remarks Populated during the pre-scan phase; used by `NRP-DOX-020`.
   */
  definedGroups: Set<string>;
  /**
   * All public function and class names declared in project headers.
   *
   * @remarks Used to validate `@see` cross-references.
   */
  definedSymbols: Set<string>;
}

/**
 * Determines the structural role of a file from its path and content.
 *
 * @param filePath - Absolute path to the file.
 * @param _raw - Raw file content (reserved for future content-based detection).
 * @returns The {@link FileRole} that controls which rule subsets apply.
 */
function detectRole(filePath: string, _raw: string): FileRole {
  const base = path.basename(filePath);
  const ext = path.extname(filePath).toLowerCase();

  if (base === "module-doc.h") return "module-doc";
  if (/(?:test[_-]|[_-]test\.|_tests?\.)/.test(base)) return "test";
  if (
    filePath.includes("third_party") ||
    filePath.includes("nordic_sdk") ||
    filePath.includes("nrf_cc310") ||
    filePath.includes("CMSIS")
  ) return "third-party";

  if (ext === ".s" || ext === ".S") return "asm";
  if (base === "CMakeLists.txt" || ext === ".cmake") return "cmake";
  if (base === "Kconfig" || base === "Doxyfile") return "config";
  if (ext === ".h" || ext === ".hpp") return "public-header";
  return "impl";
}

/**
 * Infers the target hardware platform from preprocessor includes and macros.
 *
 * @param raw - Raw file content to scan for platform markers.
 * @param configHint - Platform value from `nirapod-audit.toml`; returned
 *   unchanged unless it is `"auto"`.
 * @returns The resolved {@link PlatformHint}.
 *
 * @remarks
 * Detection order: nRF5340 markers → nRF52840 markers → ESP32 markers →
 * `"host"` fallback. The first match wins.
 */
function detectPlatform(raw: string, configHint: PlatformHint): PlatformHint {
  if (configHint !== "auto") return configHint;
  if (raw.includes("nrf5340") || raw.includes("NRF5340") || raw.includes("CC312")) return "nrf5340";
  if (raw.includes("nrf52840") || raw.includes("NRF52840") || raw.includes("CC310")) return "nrf52840";
  if (raw.includes("esp_aes.h") || raw.includes("IDF_TARGET_ESP32")) return "esp32";
  return "host";
}

/**
 * Detects the source language from a file extension.
 *
 * @param filePath - Absolute path to the source file.
 * @returns The detected {@link Language}, or `null` for unsupported files.
 *
 * @remarks
 * Extension mapping:
 * - `.c`, `.h` → `"c"`
 * - `.cpp`, `.cc`, `.hpp` → `"cpp"`
 * - `.ts`, `.tsx` → `"typescript"`
 * - `.rs` → `"rust"`
 *
 * The pipeline skips files that return `null` (config files, Makefiles, etc.).
 */
export function detectLanguage(filePath: string): Language | null {
  const ext = path.extname(filePath).toLowerCase();
  switch (ext) {
    case ".c":
    case ".h":
      return "c";
    case ".cpp":
    case ".cc":
    case ".hpp":
      return "cpp";
    case ".ts":
    case ".tsx":
      return "typescript";
    case ".rs":
      return "rust";
    default:
      return null;
  }
}

/**
 * Builds a fully-populated {@link FileContext} for one source file.
 *
 * @param filePath - Absolute path to the file to analyse.
 * @param raw - Raw file content (caller reads with `Bun.file().text()`).
 * @param project - Shared project-wide context (config, symbol table, etc.).
 * @returns A read-only context object ready to pass to each analysis pass.
 *
 * @remarks
 * Selects the C++ parser for `.cpp`, `.cc`, and `.hpp` files; falls back to
 * the C parser for everything else. Both parsers are module-level singletons.
 *
 * @throws {Error} If tree-sitter fails to parse the file (malformed source).
 *
 * @example
 * ```typescript
 * const raw = await Bun.file(filePath).text();
 * const ctx = buildFileContext(filePath, raw, projectCtx);
 * for (const pass of passes) {
 *   diagnostics.push(...pass.run(ctx));
 * }
 * ```
 */
export function buildFileContext(
  filePath: string,
  raw: string,
  project: ProjectContext,
): FileContext {
  const ext = path.extname(filePath).toLowerCase();
  const isCpp = ext === ".cpp" || ext === ".cc" || ext === ".hpp";
  const language = detectLanguage(filePath) ?? (isCpp ? "cpp" : "c");
  const parser = isCpp ? cppParser : cParser;
  const tree = parser.parse(raw);

  return {
    path: filePath,
    raw,
    lines: raw.split("\n"),
    tree,
    rootNode: tree.rootNode,
    role: detectRole(filePath, raw),
    platform: detectPlatform(raw, project.config.platform),
    language,
    isCpp,
    project,
  };
}

File: crates/core/src/context.rs
// SPDX-License-Identifier: APACHE-2.0
// SPDX-FileCopyrightText: 2026 Nirapod Contributors

//! File and project context builders for the analysis pipeline.
//!
//! Each source file is parsed once and handed to later passes through a
//! read-only [`FileContext`]. That keeps parser setup centralized and avoids
//! repeated parse work inside every rule pass.

use crate::{parse_source, AuditConfig, FileRole, ParserError, PlatformHint, SourceLanguage};
use std::path::{Path, PathBuf};
use tree_sitter::Tree;

/// Shared state built once for an audit target.
///
/// # Examples
///
/// ```
/// use nirapod_audit_core::{build_project_context, AuditConfig};
/// use std::path::PathBuf;
///
/// let project = build_project_context(".", vec![PathBuf::from("src/lib.rs")], AuditConfig::default());
/// assert_eq!(project.all_files.len(), 1);
/// ```
#[derive(Debug, Clone)]
pub struct ProjectContext {
    /// Root directory of the audit target.
    pub root_dir: PathBuf,
    /// Discovered source files in stable order.
    pub all_files: Vec<PathBuf>,
    /// Active audit configuration.
    pub config: AuditConfig,
}

/// Parsed context for one source file.
pub struct FileContext {
    /// Source path.
    pub path: PathBuf,
    /// Raw source content.
    pub raw: String,
    /// Source split into lines for later span work.
    pub lines: Vec<String>,
    /// Parsed tree-sitter tree.
    pub tree: Tree,
    /// Structural role assigned to the file.
    pub role: FileRole,
    /// Resolved platform hint for the file.
    pub platform: PlatformHint,
    /// Detected source language.
    pub language: SourceLanguage,
    /// `true` for C++ sources and headers.
    pub is_cpp: bool,
}

/// Errors returned while constructing a [`FileContext`].
#[derive(Debug)]
pub enum ContextBuildError {
    /// The file path does not map to a supported source language.
    UnsupportedPath(PathBuf),
    /// Parsing the source failed.
    Parse(ParserError),
}

impl std::fmt::Display for ContextBuildError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::UnsupportedPath(path) => write!(
                f,
                "file path does not map to a supported source language: {}",
                path.display()
            ),
            Self::Parse(error) => write!(f, "{error}"),
        }
    }
}

impl std::error::Error for ContextBuildError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            Self::Parse(error) => Some(error),
            Self::UnsupportedPath(_) => None,
        }
    }
}

impl From<ParserError> for ContextBuildError {
    fn from(error: ParserError) -> Self {
        Self::Parse(error)
    }
}

/// Builds a project context from a root directory, file list, and config.
///
/// # Examples
///
/// ```
/// use nirapod_audit_core::{build_project_context, AuditConfig};
/// use std::path::PathBuf;
///
/// let project = build_project_context("tests", vec![PathBuf::from("tests/example.h")], AuditConfig::default());
/// assert_eq!(project.root_dir, PathBuf::from("tests"));
/// ```
#[must_use]
pub fn build_project_context(
    root_dir: impl AsRef<Path>,
    all_files: Vec<PathBuf>,
    config: AuditConfig,
) -> ProjectContext {
    ProjectContext {
        root_dir: root_dir.as_ref().to_path_buf(),
        all_files,
        config,
    }
}

/// Detects a file role from its path and filename.
///
/// # Examples
///
/// ```
/// use nirapod_audit_core::{detect_file_role, FileRole};
///
/// assert_eq!(detect_file_role("include/module-doc.h"), FileRole::ModuleDoc);
/// assert_eq!(detect_file_role("src/crypto_test.cpp"), FileRole::Test);
/// ```
#[must_use]
pub fn detect_file_role(path: impl AsRef<Path>) -> FileRole {
    let path = path.as_ref();
    let file_name = path
        .file_name()
        .map(|value| value.to_string_lossy().to_ascii_lowercase())
        .unwrap_or_default();
    let normalized = path
        .to_string_lossy()
        .replace('\\', "/")
        .to_ascii_lowercase();
    let extension = path
        .extension()
        .map(|value| value.to_string_lossy().to_ascii_lowercase());

    if file_name == "module-doc.h" {
        return FileRole::ModuleDoc;
    }
    if normalized.contains("third_party/")
        || normalized.contains("third-party/")
        || normalized.contains("vendor/")
        || normalized.contains("nordic_sdk/")
    {
        return FileRole::ThirdParty;
    }
    if file_name == "cmakelists.txt" || extension.as_deref() == Some("cmake") {
        return FileRole::Cmake;
    }
    if file_name == "kconfig" || file_name == "doxyfile" {
        return FileRole::Config;
    }
    if matches!(extension.as_deref(), Some("s" | "asm")) {
        return FileRole::Asm;
    }
    if file_name.contains("test") {
        return FileRole::Test;
    }
    if matches!(extension.as_deref(), Some("h" | "hpp")) {
        return FileRole::PublicHeader;
    }
    FileRole::Implementation
}

/// Resolves a platform hint from source text and config.
///
/// The config value wins unless it is [`PlatformHint::Auto`].
///
/// # Examples
///
/// ```
/// use nirapod_audit_core::{resolve_platform_hint, PlatformHint};
///
/// let hint = resolve_platform_hint("uses CC312 here", PlatformHint::Auto);
/// assert_eq!(hint, PlatformHint::Nrf5340);
/// ```
#[must_use]
pub fn resolve_platform_hint(raw: &str, config_hint: PlatformHint) -> PlatformHint {
    if config_hint != PlatformHint::Auto {
        return config_hint;
    }
    if raw.contains("nrf5340") || raw.contains("NRF5340") || raw.contains("CC312") {
        return PlatformHint::Nrf5340;
    }
    if raw.contains("nrf52840") || raw.contains("NRF52840") || raw.contains("CC310") {
        return PlatformHint::Nrf52840;
    }
    if raw.contains("IDF_TARGET_ESP32") || raw.contains("esp_aes.h") {
        return PlatformHint::Esp32;
    }
    PlatformHint::Host
}

/// Builds a parsed file context for one supported source file.
///
/// # Errors
///
/// Returns [`ContextBuildError::UnsupportedPath`] when the file extension is
/// not supported by the current migration phase, or [`ContextBuildError::Parse`]
/// if tree-sitter parsing fails.
///
/// # Examples
///
/// ```
/// use nirapod_audit_core::{build_file_context, build_project_context, AuditConfig};
/// use std::path::PathBuf;
///
/// let project = build_project_context(".", vec![PathBuf::from("example.h")], AuditConfig::default());
/// let ctx = build_file_context("example.h", "int api(void);", &project)?;
/// assert_eq!(ctx.lines.len(), 1);
/// # Ok::<(), nirapod_audit_core::ContextBuildError>(())
/// ```
pub fn build_file_context(
    path: impl AsRef<Path>,
    raw: &str,
    project: &ProjectContext,
) -> Result<FileContext, ContextBuildError> {
    let path = path.as_ref();
    let language = crate::detect_language(path)
        .ok_or_else(|| ContextBuildError::UnsupportedPath(path.into()))?;
    let tree = parse_source(language, raw)?;

    Ok(FileContext {
        path: path.to_path_buf(),
        raw: raw.to_owned(),
        lines: raw.lines().map(str::to_owned).collect(),
        tree,
        role: detect_file_role(path),
        platform: resolve_platform_hint(raw, project.config.platform),
        language,
        is_cpp: language == SourceLanguage::Cpp,
    })
}

#[cfg(test)]
mod tests {
    use super::{
        build_file_context, build_project_context, detect_file_role, resolve_platform_hint,
        FileRole,
    };
    use crate::{AuditConfig, PlatformHint};
    use std::{fs, path::PathBuf};

    #[test]
    fn detects_common_file_roles() {
        assert_eq!(
            detect_file_role("include/module-doc.h"),
            FileRole::ModuleDoc
        );
        assert_eq!(detect_file_role("src/crypto_test.cpp"), FileRole::Test);
        assert_eq!(detect_file_role("vendor/sdk/aes.c"), FileRole::ThirdParty);
    }

    #[test]
    fn resolves_platform_from_source_markers() {
        assert_eq!(
            resolve_platform_hint("CC312 backend", PlatformHint::Auto),
            PlatformHint::Nrf5340
        );
        assert_eq!(
            resolve_platform_hint("CC310 backend", PlatformHint::Auto),
            PlatformHint::Nrf52840
        );
    }

    #[test]
    fn builds_file_context_for_fixture() {
        let fixture = fixture_path();
        let source = fs::read_to_string(&fixture).expect("failed to read compliant fixture");
        let project = build_project_context(
            fixture.parent().expect("fixture without parent"),
            vec![fixture.clone()],
            AuditConfig::default(),
        );
        let context =
            build_file_context(&fixture, &source, &project).expect("failed to build file context");

        assert_eq!(context.role, FileRole::PublicHeader);
        assert_eq!(context.platform, PlatformHint::Nrf5340);
        assert_eq!(context.lines.len(), source.lines().count());
    }

    fn fixture_path() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("../../tests/fixtures/compliant/good-header.h")
    }
}
